dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -559,69 +561,79 @@ public ListenableFuture<LoadFileResponse> load(
-                .setRetryable(true)
-                .setMessage(t.getMessage()).build());
-          }
-        }, GrpcExecutors.READER_EXECUTOR);
-        futures.add(loadFuture);
-      } catch (RejectedExecutionException ex) {
-        LOG.warn("BlockDataReaderExecutor overloaded.");
-        AlluxioRuntimeException t = AlluxioRuntimeException.from(ex);
-        errors.add(LoadFileFailure.newBuilder().setUfsStatus(status.toProto())
-            .setCode(t.getStatus().getCode().value())
-            .setRetryable(true)
-            .setMessage(t.getMessage()).build());
+          }, GrpcExecutors.READER_EXECUTOR);
+          futures.add(loadFuture);
+        } catch (RejectedExecutionException ex) {
+          LOG.warn("Load task overloaded.");
+          errors.add(LoadFailure.newBuilder().setBlock(block)
+                                    .setCode(Status.RESOURCE_EXHAUSTED.getCode().value())
+                                    .setRetryable(true).setMessage(ex.getMessage()).build());
+        }
       }
     }
-    return Futures.whenAllComplete(futures).call(() -> LoadFileResponse.newBuilder()
-            .addAllFailures(errors)
-            .setBytesSkipped(skippedFileLength.get())
-            .setFilesSkipped(skippedFiles.get())
-            // Status is a required field, put it as a placeholder
-            .setStatus(TaskStatus.SUCCESS)
-            .build(),
+    return Futures.whenAllComplete(futures).call(
+        () -> LoadFileResponse.newBuilder().addAllFailures(errors)
+                              .setBytesSkipped(skippedLength.get())
+                              .setNumSkipped(numSkipped.get())
+                              // Status is a required field, put it as a placeholder
+                              .setStatus(TaskStatus.SUCCESS).build(),
         GrpcExecutors.READER_EXECUTOR);
   }
 
-  protected void loadData(String ufsPath, long mountId, long length)
-      throws AccessControlException, IOException {
+  protected void loadData(String ufsPath, long mountId, long offset, long lengthToLoad,
+      long fileLength) throws AccessControlException, IOException {
     Protocol.OpenUfsBlockOptions options =
         Protocol.OpenUfsBlockOptions.newBuilder().setUfsPath(ufsPath).setMountId(mountId)
-            .setNoCache(false).setOffsetInFile(0).setBlockSize(length)
-            .build();
+                                    .setNoCache(false).setOffsetInFile(offset)
+                                    .setBlockSize(fileLength).build();
     String fileId = new AlluxioURI(ufsPath).hash();
-    ByteBuf buf = PooledDirectNioByteBuf.allocate((int) (4 * mPageSize));
-    try (BlockReader fileReader = createFileReader(fileId, 0, false, options)) {
-      // cache file data
-      while (fileReader.transferTo(buf) != -1) {
+    long bufferSize = 4 * mPageSize;
+    ByteBuf buf =
+        PooledByteBufAllocator.DEFAULT.directBuffer((int) Math.min(bufferSize, lengthToLoad));
```

Comment:
make it min of bufferSize or 1 page size if lengthToLoad is small, so these aligned buffers can be reused and reduce  framentation.

Reply:
done